Unable to use the object with key value 'templateUrl' in angular6

<!--
IF YOU DON'T FILL OUT THE FOLLOWING INFORMATION YOUR ISSUE MIGHT BE CLOSED WITHOUT INVESTIGATING
-->
### Bug Report or Feature Request (mark with an `x`)
```
- [x ] bug report -> please search issues before submitting
```

### Command (mark with an `x`)
```
- [ ] new
- [ ] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Versions
"@angular-devkit/build-angular": "~0.7.0",
    "@angular-devkit/build-ng-packagr": "~0.7.0",
    "@angular/cli": "~6.1.5",
    "@angular/compiler-cli": "^6.1.0",
    "@angular/language-service": "^6.1.0",
    "@types/jasmine": "~2.8.6",
    "@types/jasminewd2": "~2.0.3",
    "@types/node": "~8.9.4",
    "codelyzer": "~4.2.1",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~1.7.1",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.0",
    "karma-jasmine": "~1.1.1",
    "karma-jasmine-html-reporter": "^0.2.2",
    "ng-packagr": "^3.0.0",
    "protractor": "~5.4.0",
    "ts-node": "~5.0.1",
    "tsickle": ">=0.29.0",
    "tslib": "^1.9.0",
    "tslint": "~5.9.1",
    "typescript": "~2.7.2"
Angular CLI: 1.6.6


### Repro steps
I have the below object in component.ts and I am going to iterate this array in the HTML template inside ui-comp-menu.

menuObject = [{
'labels': 'Content1',
'templateUrl': 'assets/partials/sample.html',
'childrens': [],
},
{
'labels': 'Content2',
'templateUrl': 'assets/partials/sample1.html',
}];

and in the html

<ui-comp-menu [data]="menuObject" [label]="'labels'" [url]="'routeUrl'" >
But while compiling I got the below error




### The log given by the failure
ERROR in ./src/app/app.component.ts

Module not found: Error: Can't resolve './assets/partials/UserAdministration.html' in 'D:\POC_Angular_6\lib-demo\src\appI


### Desired functionality
It should work even if you give an object with 'templateUrl' as key name


### Mention any other details that might be useful
If i change the object key 'templateUrl' to some other name like 'templatePath' issue won't come and compiled successfully.

Similarly 'styleUrl' also not supported.

Ref. this: https://stackoverflow.com/questions/52648392/unable-to-use-the-object-with-key-value-templateurl-in-angular6/52648863#52648863

--- a/packages/ngtools/webpack/src/angular_compiler_plugin.ts
+++ b/packages/ngtools/webpack/src/angular_compiler_plugin.ts
@@ -849,7 +849,7 @@ export class AngularCompilerPlugin {
 
     if (this._JitMode) {
       // Replace resources in JIT.
-      this._transformers.push(replaceResources(isAppPath));
+      this._transformers.push(replaceResources(isAppPath, getTypeChecker));
     } else {
       // Remove unneeded angular decorators.
       this._transformers.push(removeDecorators(isAppPath, getTypeChecker));
@@ -1054,9 +1054,7 @@ export class AngularCompilerPlugin {
       .filter(x => x);
 
     const resourceImports = findResources(sourceFile)
-      .map((resourceReplacement) => resourceReplacement.resourcePaths)
-      .reduce((prev, curr) => prev.concat(curr), [])
-      .map((resourcePath) => resolve(dirname(resolvedFileName), normalize(resourcePath)));
+      .map(resourcePath => resolve(dirname(resolvedFileName), normalize(resourcePath)));
 
     // These paths are meant to be used by the loader so we must denormalize them.
     const uniqueDependencies = new Set([

--- a/packages/ngtools/webpack/src/transformers/find_resources.ts
+++ b/packages/ngtools/webpack/src/transformers/find_resources.ts
@@ -1,0 +1,68 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as ts from 'typescript';
+import { collectDeepNodes } from './ast_helpers';
+import { getResourceUrl } from './replace_resources';
+
+export function findResources(sourceFile: ts.SourceFile): string[] {
+  const resources: string[] = [];
+  const decorators = collectDeepNodes<ts.Decorator>(sourceFile, ts.SyntaxKind.Decorator);
+
+  for (const node of decorators) {
+    if (!ts.isCallExpression(node.expression)) {
+      continue;
+    }
+
+    const decoratorFactory = node.expression;
+    const args = decoratorFactory.arguments;
+    if (args.length !== 1 || !ts.isObjectLiteralExpression(args[0])) {
+      // Unsupported component metadata
+      continue;
+    }
+
+    ts.visitNodes(
+      (args[0] as ts.ObjectLiteralExpression).properties,
+      (node: ts.ObjectLiteralElementLike) => {
+        if (!ts.isPropertyAssignment(node) || ts.isComputedPropertyName(node.name)) {
+          return node;
+        }
+
+        const name = node.name.text;
+        switch (name) {
+          case 'templateUrl':
+            const url = getResourceUrl(node.initializer);
+
+            if (url) {
+              resources.push(url);
+            }
+            break;
+
+          case 'styleUrls':
+            if (!ts.isArrayLiteralExpression(node.initializer)) {
+              return node;
+            }
+
+            ts.visitNodes(node.initializer.elements, (node: ts.Expression) => {
+              const url = getResourceUrl(node);
+
+              if (url) {
+                resources.push(url);
+              }
+
+              return node;
+            });
+            break;
+        }
+
+        return node;
+      },
+    );
+  }
+
+  return resources;
+}

--- a/packages/ngtools/webpack/src/transformers/find_resources_spec.ts
+++ b/packages/ngtools/webpack/src/transformers/find_resources_spec.ts
@@ -1,0 +1,49 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { tags } from '@angular-devkit/core';  // tslint:disable-line:no-implicit-dependencies
+import * as ts from 'typescript';
+import { findResources } from './find_resources';
+
+
+describe('@ngtools/webpack transformers', () => {
+  describe('find_resources', () => {
+    it('should return resources', () => {
+      const input = tags.stripIndent`
+        import { Component } from '@angular/core';
+
+        @Component({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css', './app.component.2.css']
+        })
+        export class AppComponent {
+          title = 'app';
+        }
+      `;
+
+      const result = findResources(ts.createSourceFile('temp.ts', input, ts.ScriptTarget.ES2015));
+      expect(result).toEqual([
+        './app.component.html',
+        './app.component.css',
+        './app.component.2.css',
+      ]);
+    });
+
+    it('should not return resources if they are not in decorator', () => {
+      const input = tags.stripIndent`
+        const foo = {
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css', './app.component.2.css']
+        }
+      `;
+
+      const result = findResources(ts.createSourceFile('temp.ts', input, ts.ScriptTarget.ES2015));
+      expect(result).toEqual([]);
+    });
+  });
+});

--- a/packages/ngtools/webpack/src/transformers/index.ts
+++ b/packages/ngtools/webpack/src/transformers/index.ts
@@ -17,3 +17,4 @@ export * from './export_lazy_module_map';
 export * from './register_locale_data';
 export * from './replace_resources';
 export * from './remove_decorators';
+export * from './find_resources';

--- a/packages/ngtools/webpack/src/transformers/replace_resources.ts
+++ b/packages/ngtools/webpack/src/transformers/replace_resources.ts
@@ -6,155 +6,214 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import * as ts from 'typescript';
-import { collectDeepNodes, getFirstNode } from './ast_helpers';
-import {
-  AddNodeOperation,
-  ReplaceNodeOperation,
-  StandardTransform,
-  TransformOperation,
-} from './interfaces';
-import { makeTransform } from './make_transform';
-
 
 export function replaceResources(
   shouldTransform: (fileName: string) => boolean,
+  getTypeChecker: () => ts.TypeChecker,
 ): ts.TransformerFactory<ts.SourceFile> {
-  const standardTransform: StandardTransform = function (sourceFile: ts.SourceFile) {
-    const ops: TransformOperation[] = [];
 
-    if (!shouldTransform(sourceFile.fileName)) {
-      return ops;
-    }
+  return (context: ts.TransformationContext) => {
+    const typeChecker = getTypeChecker();
 
-    const replacements = findResources(sourceFile);
+    const visitNode: ts.Visitor = (node: ts.Decorator) => {
+      if (ts.isClassDeclaration(node)) {
+        node.decorators = ts.visitNodes(
+          node.decorators,
+          (node: ts.Decorator) => visitDecorator(node, typeChecker),
+        );
+      }
 
-    if (replacements.length > 0) {
+      return ts.visitEachChild(node, visitNode, context);
+    };
 
-      // Add the replacement operations.
-      ops.push(...(replacements.map((rep) => rep.replaceNodeOperation)));
+    return (sourceFile: ts.SourceFile) => (
+      shouldTransform(sourceFile.fileName)
+        ? ts.visitNode(sourceFile, visitNode)
+        : sourceFile
+    );
+  };
+}
 
-      // If we added a require call, we need to also add typings for it.
-      // The typings need to be compatible with node typings, but also work by themselves.
+function visitDecorator(node: ts.Decorator, typeChecker: ts.TypeChecker): ts.Decorator {
+  if (!isComponentDecorator(node, typeChecker)) {
+    return node;
+  }
 
-      // interface NodeRequire {(id: string): any;}
-      const nodeRequireInterface = ts.createInterfaceDeclaration([], [], 'NodeRequire', [], [], [
-        ts.createCallSignature([], [
-          ts.createParameter([], [], undefined, 'id', undefined,
-            ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
-          ),
-        ], ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)),
-      ]);
+  if (!ts.isCallExpression(node.expression)) {
+    return node;
+  }
 
-      // declare var require: NodeRequire;
-      const varRequire = ts.createVariableStatement(
-        [ts.createToken(ts.SyntaxKind.DeclareKeyword)],
-        [ts.createVariableDeclaration('require', ts.createTypeReferenceNode('NodeRequire', []))],
+  const decoratorFactory = node.expression;
+  const args = decoratorFactory.arguments;
+  if (args.length !== 1 || !ts.isObjectLiteralExpression(args[0])) {
+    // Unsupported component metadata
+    return node;
+  }
+
+  const objectExpression = args[0] as ts.ObjectLiteralExpression;
+  const styleReplacements: ts.Expression[] = [];
+
+  // visit all properties
+  let properties = ts.visitNodes(
+    objectExpression.properties,
+    (node: ts.ObjectLiteralElementLike) => visitComponentMetadata(node, styleReplacements),
+  );
+
+  // replace properties with updated properties
+  if (styleReplacements.length > 0) {
+    const styleProperty = ts.createPropertyAssignment(
+      ts.createIdentifier('styles'),
+      ts.createArrayLiteral(styleReplacements),
+    );
+
+    properties = ts.createNodeArray([...properties, styleProperty]);
+  }
+
+  return ts.updateDecorator(
+    node,
+    ts.updateCall(
+      decoratorFactory,
+      decoratorFactory.expression,
+      decoratorFactory.typeArguments,
+      [ts.updateObjectLiteral(objectExpression, properties)],
+    ),
+  );
+}
+
+function visitComponentMetadata(
+  node: ts.ObjectLiteralElementLike,
+  styleReplacements: ts.Expression[],
+): ts.ObjectLiteralElementLike | undefined {
+  if (!ts.isPropertyAssignment(node) || ts.isComputedPropertyName(node.name)) {
+    return node;
+  }
+
+  const name = node.name.text;
+  switch (name) {
+    case 'moduleId':
+
+      return undefined;
+
+    case 'templateUrl':
+      return ts.updatePropertyAssignment(
+        node,
+        ts.createIdentifier('template'),
+        createRequireExpression(node.initializer),
       );
 
-      ops.push(new AddNodeOperation(sourceFile, getFirstNode(sourceFile), nodeRequireInterface));
-      ops.push(new AddNodeOperation(sourceFile, getFirstNode(sourceFile), varRequire));
-    }
-
-    return ops;
-  };
-
-  return makeTransform(standardTransform);
-}
-
-export interface ResourceReplacement {
-  resourcePaths: string[];
-  replaceNodeOperation: ReplaceNodeOperation;
-}
-
-export function findResources(sourceFile: ts.SourceFile): ResourceReplacement[] {
-  const replacements: ResourceReplacement[] = [];
-
-  // Find all object literals.
-  collectDeepNodes<ts.ObjectLiteralExpression>(sourceFile, ts.SyntaxKind.ObjectLiteralExpression)
-    // Get all their property assignments.
-    .map(node => collectDeepNodes<ts.PropertyAssignment>(node, ts.SyntaxKind.PropertyAssignment))
-    // Flatten into a single array (from an array of array<property assignments>).
-    .reduce((prev, curr) => curr ? prev.concat(curr) : prev, [])
-    // We only want property assignments for the templateUrl/styleUrls keys.
-    .filter((node: ts.PropertyAssignment) => {
-      const key = _getContentOfKeyLiteral(node.name);
-      if (!key) {
-        // key is an expression, can't do anything.
-        return false;
+    case 'styles':
+    case 'styleUrls':
+      if (!ts.isArrayLiteralExpression(node.initializer)) {
+        return node;
       }
 
-      return key == 'templateUrl' || key == 'styleUrls';
-    })
-    // Replace templateUrl/styleUrls key with template/styles, and and paths with require('path').
-    .forEach((node: ts.PropertyAssignment) => {
-      const key = _getContentOfKeyLiteral(node.name);
+      const isInlineStyles = name === 'styles';
+      const styles = ts.visitNodes(
+        node.initializer.elements,
+        (node: ts.Expression) => {
+          if (!ts.isStringLiteral(node) && !ts.isNoSubstitutionTemplateLiteral(node)) {
+            return node;
+          }
 
-      if (key == 'templateUrl') {
-        const resourcePath = _getResourceRequest(node.initializer, sourceFile);
-        const requireCall = _createRequireCall(resourcePath);
-        const propAssign = ts.createPropertyAssignment('template', requireCall);
-        replacements.push({
-          resourcePaths: [resourcePath],
-          replaceNodeOperation: new ReplaceNodeOperation(sourceFile, node, propAssign),
-        });
-      } else if (key == 'styleUrls') {
-        const arr = collectDeepNodes<ts.ArrayLiteralExpression>(node,
-          ts.SyntaxKind.ArrayLiteralExpression);
-        if (!arr || arr.length == 0 || arr[0].elements.length == 0) {
-          return;
-        }
+          return isInlineStyles ? ts.createLiteral(node.text) : createRequireExpression(node);
+        },
+      );
 
-        const stylePaths = arr[0].elements.map((element: ts.Expression) => {
-          return _getResourceRequest(element, sourceFile);
-        });
-
-        const requireArray = ts.createArrayLiteral(
-          stylePaths.map((path) => _createRequireCall(path)),
-        );
-
-        const propAssign = ts.createPropertyAssignment('styles', requireArray);
-        replacements.push({
-          resourcePaths: stylePaths,
-          replaceNodeOperation: new ReplaceNodeOperation(sourceFile, node, propAssign),
-        });
+      // Styles should be placed first
+      if (isInlineStyles) {
+        styleReplacements.unshift(...styles);
+      } else {
+        styleReplacements.push(...styles);
       }
-    });
 
-  return replacements;
+      return undefined;
 
-}
-
-function _getContentOfKeyLiteral(node?: ts.Node): string | null {
-  if (!node) {
-    return null;
-  } else if (node.kind == ts.SyntaxKind.Identifier) {
-    return (node as ts.Identifier).text;
-  } else if (node.kind == ts.SyntaxKind.StringLiteral) {
-    return (node as ts.StringLiteral).text;
-  } else {
-    return null;
+    default:
+      return node;
   }
 }
 
-function _getResourceRequest(element: ts.Expression, sourceFile: ts.SourceFile) {
-  if (
-    element.kind === ts.SyntaxKind.StringLiteral ||
-    element.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral
-  ) {
-    const url = (element as ts.StringLiteral).text;
-
-    // If the URL does not start with ./ or ../, prepends ./ to it.
-    return `${/^\.?\.\//.test(url) ? '' : './'}${url}`;
-  } else {
-    // if not string, just use expression directly
-    return element.getFullText(sourceFile);
+export function getResourceUrl(node: ts.Expression): string | null {
+  // only analyze strings
+  if (!ts.isStringLiteral(node) && !ts.isNoSubstitutionTemplateLiteral(node)) {
+    return null;
   }
+
+  return `${/^\.?\.\//.test(node.text) ? '' : './'}${node.text}`;
 }
 
-function _createRequireCall(path: string) {
+function isComponentDecorator(node: ts.Node, typeChecker: ts.TypeChecker): node is ts.Decorator {
+  if (!ts.isDecorator(node)) {
+    return false;
+  }
+
+  const origin = getDecoratorOrigin(node, typeChecker);
+  if (origin && origin.module === '@angular/core' && origin.name === 'Component') {
+    return true;
+  }
+
+  return false;
+}
+
+function createRequireExpression(node: ts.Expression): ts.Expression {
+  const url = getResourceUrl(node);
+  if (!url) {
+    return node;
+  }
+
   return ts.createCall(
     ts.createIdentifier('require'),
-    [],
-    [ts.createLiteral(path)],
+    undefined,
+    [ts.createLiteral(url)],
   );
+}
+
+interface DecoratorOrigin {
+  name: string;
+  module: string;
+}
+
+function getDecoratorOrigin(
+  decorator: ts.Decorator,
+  typeChecker: ts.TypeChecker,
+): DecoratorOrigin | null {
+  if (!ts.isCallExpression(decorator.expression)) {
+    return null;
+  }
+
+  let identifier: ts.Node;
+  let name = '';
+
+  if (ts.isPropertyAccessExpression(decorator.expression.expression)) {
+    identifier = decorator.expression.expression.expression;
+    name = decorator.expression.expression.name.text;
+  } else if (ts.isIdentifier(decorator.expression.expression)) {
+    identifier = decorator.expression.expression;
+  } else {
+    return null;
+  }
+
+  // NOTE: resolver.getReferencedImportDeclaration would work as well but is internal
+  const symbol = typeChecker.getSymbolAtLocation(identifier);
+  if (symbol && symbol.declarations && symbol.declarations.length > 0) {
+    const declaration = symbol.declarations[0];
+    let module: string;
+
+    if (ts.isImportSpecifier(declaration)) {
+      name = (declaration.propertyName || declaration.name).text;
+      module = (declaration.parent.parent.parent.moduleSpecifier as ts.Identifier).text;
+    } else if (ts.isNamespaceImport(declaration)) {
+      // Use the name from the decorator namespace property access
+      module = (declaration.parent.parent.moduleSpecifier as ts.Identifier).text;
+    } else if (ts.isImportClause(declaration)) {
+      name = (declaration.name as ts.Identifier).text;
+      module = (declaration.parent.moduleSpecifier as ts.Identifier).text;
+    } else {
+      return null;
+    }
+
+    return { name, module };
+  }
+
+  return null;
 }

--- a/packages/ngtools/webpack/src/transformers/replace_resources_spec.ts
+++ b/packages/ngtools/webpack/src/transformers/replace_resources_spec.ts
@@ -6,11 +6,21 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import { tags } from '@angular-devkit/core';  // tslint:disable-line:no-implicit-dependencies
-import { transformTypescript } from './ast_helpers';
+import { createTypescriptContext, transformTypescript } from './ast_helpers';
 import { replaceResources } from './replace_resources';
 
+function transform(input: string, shouldTransform = true) {
+  const { program } = createTypescriptContext(input);
+  const getTypeChecker = () => program.getTypeChecker();
+  const transformer = replaceResources(() => shouldTransform, getTypeChecker);
+
+  return transformTypescript(input, [transformer]);
+}
+
+// tslint:disable-next-line:no-big-function
 describe('@ngtools/webpack transformers', () => {
-  describe('replace_resources', () => {
+  // tslint:disable-next-line:no-big-function
+  describe('find_resources', () => {
     it('should replace resources', () => {
       const input = tags.stripIndent`
         import { Component } from '@angular/core';
@@ -42,9 +52,43 @@ describe('@ngtools/webpack transformers', () => {
         export { AppComponent };
       `;
 
-      const transformer = replaceResources(() => true);
-      const result = transformTypescript(input, [transformer]);
+      const result = transform(input);
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
 
+    it('should merge styleUrls with styles', () => {
+      const input = tags.stripIndent`
+        import { Component } from '@angular/core';
+
+        @Component({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styles: ['a { color: red }'],
+          styleUrls: ['./app.component.css'],
+        })
+        export class AppComponent {
+          title = 'app';
+        }
+      `;
+      const output = tags.stripIndent`
+        import * as tslib_1 from "tslib";
+        import { Component } from '@angular/core';
+        let AppComponent = class AppComponent {
+            constructor() {
+                this.title = 'app';
+            }
+        };
+        AppComponent = tslib_1.__decorate([
+            Component({
+                selector: 'app-root',
+                template: require("./app.component.html"),
+                styles: ["a { color: red }", require("./app.component.css")]
+            })
+        ], AppComponent);
+        export { AppComponent };
+      `;
+
+      const result = transform(input);
       expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
     });
 
@@ -79,9 +123,177 @@ describe('@ngtools/webpack transformers', () => {
         export { AppComponent };
       `;
 
-      const transformer = replaceResources(() => true);
+      const result = transform(input);
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+    it('should replace resources if Component decorator is aliased', () => {
+      const input = tags.stripIndent`
+        import { Component as NgComponent } from '@angular/core';
+
+        @NgComponent({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css', './app.component.2.css']
+        })
+        export class AppComponent {
+          title = 'app';
+        }
+      `;
+      const output = tags.stripIndent`
+        import * as tslib_1 from "tslib";
+        import { Component as NgComponent } from '@angular/core';
+        let AppComponent = class AppComponent {
+            constructor() {
+                this.title = 'app';
+            }
+        };
+        AppComponent = tslib_1.__decorate([
+          NgComponent({
+                selector: 'app-root',
+                template: require("./app.component.html"),
+                styles: [require("./app.component.css"), require("./app.component.2.css")]
+            })
+        ], AppComponent);
+        export { AppComponent };
+      `;
+
+      const { program } = createTypescriptContext(input);
+      const getTypeChecker = () => program.getTypeChecker();
+      const transformer = replaceResources(() => true, getTypeChecker);
       const result = transformTypescript(input, [transformer]);
 
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+    it('should replace resources if Angular Core import is namespaced', () => {
+      const input = tags.stripIndent`
+        import * as ng from '@angular/core';
+
+        @ng.Component({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css', './app.component.2.css']
+        })
+        export class AppComponent {
+          title = 'app';
+        }
+      `;
+      const output = tags.stripIndent`
+        import * as tslib_1 from "tslib";
+        import * as ng from '@angular/core';
+        let AppComponent = class AppComponent {
+            constructor() {
+                this.title = 'app';
+            }
+        };
+        AppComponent = tslib_1.__decorate([
+          ng.Component({
+                selector: 'app-root',
+                template: require("./app.component.html"),
+                styles: [require("./app.component.css"), require("./app.component.2.css")]
+            })
+        ], AppComponent);
+        export { AppComponent };
+      `;
+
+      const result = transform(input);
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+    it('should not replace resources if not in Component decorator', () => {
+      const input = tags.stripIndent`
+        import { Component } from '@angular/core';
+
+        @Component({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css']
+        })
+        export class AppComponent {
+          obj = [
+            {
+              'labels': 'Content',
+              'templateUrl': 'content.html'
+            }
+          ];
+        }
+      `;
+
+      const output = tags.stripIndent`
+        import * as tslib_1 from "tslib";
+        import { Component } from '@angular/core';
+
+        let AppComponent = class AppComponent {
+          constructor() {
+            this.obj = [
+              {
+                'labels': 'Content',
+                'templateUrl': 'content.html'
+              }
+            ];
+          }
+        };
+
+        AppComponent = tslib_1.__decorate([
+            Component({
+                selector: 'app-root',
+                template: require("./app.component.html"),
+                styles: [require("./app.component.css")]
+            })
+        ], AppComponent);
+        export { AppComponent };
+      `;
+
+      const result = transform(input);
+      expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
+    });
+
+    it('should not replace resources if not in an NG Component decorator', () => {
+      const input = tags.stripIndent`
+        import { Component } from 'foo';
+
+        @Component({
+          selector: 'app-root',
+          templateUrl: './app.component.html',
+          styleUrls: ['./app.component.css']
+        })
+        export class AppComponent {
+          obj = [
+            {
+              'labels': 'Content',
+              'templateUrl': 'content.html'
+            }
+          ];
+        }
+      `;
+
+      const output = tags.stripIndent`
+        import * as tslib_1 from "tslib";
+        import { Component } from 'foo';
+
+        let AppComponent = class AppComponent {
+          constructor() {
+            this.obj = [
+              {
+                'labels': 'Content',
+                'templateUrl': 'content.html'
+              }
+            ];
+          }
+        };
+
+        AppComponent = tslib_1.__decorate([
+            Component({
+                selector: 'app-root',
+                templateUrl: './app.component.html',
+                styleUrls: ['./app.component.css']
+            })
+        ], AppComponent);
+        export { AppComponent };
+      `;
+
+      const result = transform(input);
       expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
     });
 
@@ -116,9 +328,7 @@ describe('@ngtools/webpack transformers', () => {
         export { AppComponent };
       `;
 
-      const transformer = replaceResources(() => false);
-      const result = transformTypescript(input, [transformer]);
-
+      const result = transform(input, false);
       expect(tags.oneLine`${result}`).toEqual(tags.oneLine`${output}`);
     });
   });

